src/casp/services/api/services/cell_operations_service.py

```diff
@@ -95,23 +99,90 @@ def get_knn_matches(self, embeddings: np.array, model_name: str) -> t.List[t.Lis
         """
         index, index_endpoint_client = self.__get_match_index_endpoint_client_for_model(model_name=model_name)
 
+        # Break embeddings into chunks so we don't overload the matching engine
+        num_chunks: int = math.ceil(len(embeddings) / GET_MATCHES_CHUNK_SIZE)
+        embeddings_chunks = np.array_split(embeddings, num_chunks)
+
+        all_matches = []
+        for i in range(0, num_chunks):
+            # Set up retry logic for for the matching engine requests
+            retryer = Retrying(
+                stop=stop_after_attempt(4),
+                wait=wait_exponential(multiplier=2, min=0, max=30) + wait_random(0, 2),
+                reraise=True,
+            )
+            matches = retryer(
+                self.__get_knn_matches_for_chunk,
+                embeddings_chunk=embeddings_chunks[i],
+                index=index,
+                index_endpoint_client=index_endpoint_client,
+            )
+            all_matches.extend(matches)
+
+        return all_matches
+
+    def __get_knn_matches_for_chunk(
+        self,
+        embeddings_chunk: np.array,
+        index: models.CASMatchingEngineIndex,
+        index_endpoint_client: clients.CustomMatchingEngineIndexEndpointClient,
+    ) -> t.List[t.List[MatchNeighbor]]:
+        """
+        Get KNN matches for a chunk of embeddings (split out from get_knn_matches so it can be
+        retried and called in chunks).
+
+        :param embeddings_chunk: Chunk of embeddings to match.
+        :param index: Matching engine index to use for matching.
+        :param index_endpoint_client: Matching engine client that will handle communicating with
+        the matching engine.
+
+        :return: List of lists of MatchNeighbor objects.
+        """
         matches = index_endpoint_client.match(
             deployed_index_id=index.deployed_index_id,
-            queries=embeddings,
+            queries=embeddings_chunk,
             num_neighbors=index.num_neighbors,
         )
-        self.__validate_knn_response(embeddings=embeddings, knn_response=matches)
+        self.__validate_knn_response(embeddings=embeddings_chunk, knn_response=matches)
         return matches
 
     def get_knn_matches_as_dict(self, embeddings: np.array, model_name: str) -> t.List[t.List[t.Dict[str, t.Any]]]:
         index, index_endpoint_client = self.__get_match_index_endpoint_client_for_model(model_name=model_name)
 
+        # Break embeddings into chunks so we don't overload the matching engine
+        num_chunks: int = math.ceil(len(embeddings) / GET_MATCHES_CHUNK_SIZE)
+        embeddings_chunks = np.array_split(embeddings, num_chunks)
+
+        all_matches = []
+        for i in range(0, num_chunks):
```

Comment:
Seems like this loop is identical to the one that's in `get_knn_matches` except for the method that's used to called inside of the retryer. 
Do you guys think it would be a good idea to wrap it up in some kind of a method that takes callback and its arguments (and maybe num_attempts)?

Reply:
I consolidated it into one function as you suggestion.  I changed two existing functions to wrappers that just pass the correct callback